service/src/main/java/bio/terra/externalcreds/services/ProviderService.java

```diff
@@ -122,6 +128,52 @@ public LinkedAccountWithPassportAndVisas createLink(
     return linkedAccountService.saveLinkedAccount(extractPassport(userInfo, linkedAccount));
   }
 
+  public void deleteLink(String userId, String providerId) {
+    ProviderConfig.ProviderInfo providerInfo = providerConfig.getServices().get(providerId);
+
+    if (providerInfo == null) {
+      throw new NotFoundException(String.format("Provider %s not found", providerId));
+    }
+
+    var linkedAccount =
+        linkedAccountService
+            .getLinkedAccount(userId, providerId)
+            .orElseThrow(() -> new NotFoundException("Link not found for user"));
+
+    revokeAccessToken(providerInfo, linkedAccount);
+
+    linkedAccountService.deleteLinkedAccount(userId, providerId);
+  }
+
+  private void revokeAccessToken(ProviderInfo providerInfo, LinkedAccount linkedAccount) {
+    // Get the endpoint URL and insert the token
+    String revokeEndpoint =
+        String.format(providerInfo.getRevokeEndpoint(), linkedAccount.getRefreshToken());
+    // Add authorization information and make request
+    WebClient.ResponseSpec response =
+        WebClient.create(revokeEndpoint)
+            .post()
+            .uri(
+                uriBuilder ->
+                    uriBuilder
+                        .queryParam("client_id", providerInfo.getClientId())
+                        .queryParam("client_secret", providerInfo.getClientSecret())
+                        .build())
+            .retrieve();
+
+    String responseBody =
+        response
+            .onStatus(HttpStatus::isError, clientResponse -> Mono.empty())
+            .bodyToMono(String.class)
+            .block(Duration.of(1000, ChronoUnit.MILLIS));
```

Comment:
Two quick questions about this part:
1. Is this the correct duration we should wait when calling the revoke endpoint? I had chosen 1 second arbitrarily
2. Should we throw an exception when the revoke endpoint returns an error? My thought had been that we would want to prevent the record from being deleted in the DB whenever we've failed to revoke it. Does that sound right?

Reply:
1s sounds fine to me. The revoke call should be a best effort attempt and should not prevent the user from deleting their link. So I did not throw an error on failure.